### DISCUSSION TITLE
Remove dependency on XCTest

### DISCUSSION THREAD

**Discussion by asiliuk** [Quick/Nimble#986] [Ideas]
The only file that imports `XCTest` is `NimbleXCTestHandler` and it makes it impossible to use `Nimble` outside of the test targets or targets that are only imported by tests.

When application is split in feature modules it is very handy to declare custom feature related predicates nearby the feature code. For example in Analytics module it could be nice to declare predicates that simplify assertions
```swift
// in `AnalyticsLibrary` module
protocol AnalyticsSupportingType { ... }

#if DEBUG
func haveAnalytics(...) -> Predicate<AnalyticsSupportingType>
#endif

// in `AnalyticsLibraryTests` module
expect(myType).to(haveAnalytics(...))
``` 
To do so it needs to be able `import Nimble`. So this function should either live in `AnalyticsLibraryTests` that makes it impossible to reuse it in any other test module. Or it should be declared in new module `AnalyticsLibraryTestUtils` that could only be imported by test modules.

Going the way `<Module>TestUtils` is not most convenient thing to do. First it results in x2 number of modules. And second it result in a lot of issues with linking. One example is when such module is linked to application test target that links to `host` application. In such case `AnalyticsLibrary` is linked twice, once by application module and once by test module through `AnalyticsLibraryTestUtils`.

So my question is: Have you considered a possibility to remove dependency on `XCTest` so predicate code could be declared in feature module?
1) This could be achieved by using such libraries as [pointfreeco/xctest-dynamic-overlay](https://github.com/pointfreeco/xctest-dynamic-overlay). 
2) Or maybe by using `#if canImport(XCTest)` but I'm not sure what side effects such approach will have

**Comment by younata** [comment]
Heya @asiliuk! Thanks for the feedback. My apologies for missing this until now.

We have documentation on [using nimble without xctest](https://github.com/Quick/Nimble#using-nimble-without-xctest). It's a bit of a clunky setup, to be sure. And it doesn't solve your issue with `NimbleXCTestHandler`.

I did just try `#if canImport(XCTest)` on my local checkout of Nimble and it *seems* to work just fine running the tests. I put up a [PR doing just this](https://github.com/Quick/Nimble/pull/990), I encourage you to try out that branch and let me know if it works for you. Note that you'll still need to follow the instructions around writing your own `AssertionHandler` and setting `NimbleAssertionHandler` to that, and likely you'll need to also follow the instructions of removing the `libswiftXCTest.dylib` as well.

Let me know if it works for you!

**Comment by asiliuk** [comment]
Hey @younata! Thanks for the response and opening PR!
I've checked it for my usecase and looks like it does not solve the issue 😢 I got runtime error
```
dyld[50250]: Library not loaded: @rpath/libXCTestSwiftSupport.dylib
  Referenced from: ...
```

Here's simple swift package that could reproduce this problem. I've tried with and without build post-action it does not solve the issue
[TestNimbleXCTestImport.zip](https://github.com/Quick/Nimble/files/9245108/TestNimbleXCTestImport.zip)

--- 
I've experimented on local Nimble copy and used https://github.com/pointfreeco/xctest-dynamic-overlay and it solved this problem
The only downside is that it requires macOS v10.15 and iOS v13 as minimum target version. And looks like even after [Nimble v11](https://github.com/Quick/Nimble/issues/984) it can't be used

I don't understand why it still tries to import XCTest despite `#if canImport(XCTest)` check, and I don't have enough time to try to make it work without using `xctest-dynamic-overlay`. So feel free to close PR and ignore this feature request :) I just wanted to discuss such possibility and see, maybe there are other people who would like to import Nimble in non-test code

Thanks for spending your time on this!

**Reply by younata** [reply]
> So feel free to close PR and ignore this feature request :) I just wanted to discuss such possibility and see, maybe there are other people who would like to import Nimble in non-test code.

I closed the PR. I'm going to make a proper issue for this. Because we do have documentation for how to do it, but it's either outdated or simply wrong. We should support this.
